app/src/main/java/fr/neamar/kiss/DataHandler.java

```diff
@@ -199,6 +214,46 @@ public void addToHistory(Context context, String id) {
         DBHelper.insertHistory(context, currentQuery, id);
     }
 
+    public void addShortcut(Context context, ShortcutPojo pojo) {
+        ShortcutRecord record = new ShortcutRecord();
+        record.name = pojo.name;
+        record.iconResource = pojo.resourceName;
+        record.packageName = pojo.packageName;
+        record.intentUri = pojo.intentUri;
+        
+        if (pojo.icon != null) {
+            ByteBuffer bb = ByteBuffer.allocate(pojo.icon.getRowBytes() * pojo.icon.getHeight());        
+            pojo.icon.copyPixelsToBuffer(bb);        
+            record.icon_blob = bb.array();
+        }
+
+
+        DBHelper.insertShortcut(context, record);
+    }
+
+    public ArrayList<ShortcutPojo> getShortcuts(Context context) {
+        ArrayList<ShortcutRecord> records = DBHelper.getShortcuts(context);
+        ArrayList<ShortcutPojo> pojos = new ArrayList<>();
+        for (ShortcutRecord shortcutRecord : records) {
+            ShortcutPojo pojo = getShortcutProvider().createPojo(shortcutRecord.name);
+            pojo.packageName = shortcutRecord.packageName;
+            pojo.resourceName = shortcutRecord.iconResource;
+            pojo.intentUri = shortcutRecord.intentUri;
+            if (shortcutRecord.icon_blob != null)
+                pojo.icon = BitmapFactory.decodeByteArray(shortcutRecord.icon_blob, 0, shortcutRecord.icon_blob.length);
+            
+
+            pojos.add(pojo);
+        }
+
+        return pojos;
+    }
+    
+    public void removeShortcut(Context context, String name) {
```

Comment:
idem, probably best in UninstallShorcutHandler. Since it's a one-liner, can be repeated on the boradcast and on the menu item, rather than in the DataHandler class